### PR TITLE
bpo-42179: Clarify exception chaining

### PR DESCRIPTION
**Pull request by greatvovan**
Add an important detail and provide a link.

https://bugs.python.org/issue42179

### CODE REVIEW

**Comment by merwok** [comment]
I agree with @methane that this degree of detail is too much for the tutorial.

**Comment by greatvovan** [comment]
@merwok can you post (and elaborate) your opinion in the BPO ticket? I think it's better to keep discussion on the change in one place. I can reduce degree of detail, too.

File: Doc/tutorial/errors.rst [hunk 1]
```diff
@@ -301,8 +301,13 @@ This can be useful when you are transforming exceptions. For example::
 
 The expression following the :keyword:`from` must be either an exception or
 ``None``. Exception chaining happens automatically when an exception is raised
-inside an exception handler or :keyword:`finally` section. Exception chaining
-can be disabled by using ``from None`` idiom:
+inside an exception handler or :keyword:`finally` section, however in this case
+``__context__`` attribute of the exception is used. Chaining with
```
Comment (willingc) [inline Doc/tutorial/errors.rst]: I recommend submitting this sentence right before the builtin exceptions sentence after #23162 lands. Then add "Chaining with ... takes priority."

**Review by willingc — CHANGES_REQUESTED** [review]
Thanks @greatvovan for the PR. Let's update this after #23162 lands. I like the sentence that I've flagged and think it would be helpful.

**Comment by greatvovan** [comment]
I have made the requested changes; please review again.

@willingc the sentence you suggested could not be used as is in the new variant of text as the context is bit lost after removing mentions of special attributes. I added a new abstract in the end of the section, please let me know what you think.

File: Doc/tutorial/errors.rst [hunk 2]
```diff
@@ -300,7 +300,7 @@ This can be useful when you are transforming exceptions. For example::
     RuntimeError: Failed to open database
 
 Exception chaining happens automatically when an exception is raised inside an
-:keyword:`except` or :keyword:`finally` section. Exception chaining can be
```
Comment (greatvovan) [inline Doc/tutorial/errors.rst:303]: Tautology.

File: Doc/tutorial/errors.rst [hunk 3]
```diff
@@ -281,17 +281,17 @@ chaining exceptions. For example::
 This can be useful when you are transforming exceptions. For example::
 
     >>> def func():
-    ...     raise IOError
+    ...     raise ConnectionError
```
Comment (merwok) [inline Doc/tutorial/errors.rst:284]: Nice!  One OSError was missed on line 294

File: Doc/tutorial/errors.rst [hunk 4]
```diff
@@ -312,6 +312,13 @@ disabled by using ``from None`` idiom:
       File "<stdin>", line 4, in <module>
     RuntimeError
 
+There are two special attributes that enable chaining mechanics in
+exceptions: ``__context__`` and ``__cause__``. When you raise an exception
+inside :keyword:`except` or :keyword:`finally`, Pyhton automatically fills
+``__context__`` attribute of the new exception with the caught one. When you
+use ``from exc`` syntax, Pyhon fills ``__cause__`` attribute with
+the ``exc``. You can use both attributes for introspection purposes, however
+when formatting exception traceback message, ``__cause__`` takes priority.
```
Comment (merwok) [inline Doc/tutorial/errors.rst]: For the record, I still disagree with this level of implementation notes in the tutorial.

Comment (greatvovan) [reply Doc/tutorial/errors.rst]: Removed.

**Review by willingc — APPROVED** [review]

**Comment by greatvovan** [comment]
Please review the final version.

I removed my proposal paragraph about __cause__ and __context__ as nobody seems to be happy with it. I kept only changes about exception type (https://bugs.python.org/issue42179#msg380435).

**Review by willingc — APPROVED** [review]
Thanks @greatvovan and good catch on ConnectionError. @merwok If you are good with this as it stands, please merge. Thanks :sunny:

**Comment by merwok** [comment]
Thanks for merging Carol!

FYI the commit message is messy due to github’s poor UI.

**Comment by willingc** [comment]
Sorry about the commit message. Thanks @merwok.

**Pull request merged**